### PR TITLE
Fix deadlock, improve synchronizations.

### PR DESCRIPTION
**Pull request by danchom**
Signed-off-by: Yordan Mihaylov <j.mihailov@prosyst.com>

A dead lock appares:
```
"Rules Storage Loading Thread (Bundle 46)":
        at com.prosyst.mbs.impl.services.automation.core.provider.ModuleTypeAdminResourceBundleProvider.getModuleType(ModuleTypeAdminResourceBundleProvider.java:238)
        - waiting to lock <0x00000007ac6846a0> (a java.util.HashMap)
        at org.eclipse.smarthome.automation.core.internal.type.ModuleTypeManager.get(ModuleTypeManager.java:69)
        at org.eclipse.smarthome.automation.core.internal.type.ModuleTypeManager.get(ModuleTypeManager.java:62)
        at org.eclipse.smarthome.automation.core.internal.RuleEngine.setDefaultConfigurationValues(RuleEngine.java:1412)
        at org.eclipse.smarthome.automation.core.internal.RuleEngine.validateModules(RuleEngine.java:281)
        at org.eclipse.smarthome.automation.core.internal.RuleEngine.addRule0(RuleEngine.java:239)
        at org.eclipse.smarthome.automation.core.internal.RuleEngine.addRule(RuleEngine.java:225)
        - locked <0x00000007b0f50250> (a org.eclipse.smarthome.automation.core.internal.RuleEngine)
        at org.eclipse.smarthome.automation.core.internal.RuleEngine.addRule(RuleEngine.java:212)
        - locked <0x00000007b0f50250> (a org.eclipse.smarthome.automation.core.internal.RuleEngine)
        at com.prosyst.mbs.impl.services.automation.RuleAdminImpl.run(RuleAdminImpl.java:335)
        at com.prosyst.util.impl.tpt.threadpool.ExecutorImpl.run(ExecutorImpl.java:200)
"Resource Bundle Module Types Storage Loading Thread (Bundle 53)":
        at org.eclipse.smarthome.automation.core.internal.RuleEngine.getRuleStatus(RuleEngine.java:1183)
        - waiting to lock <0x00000007b0f50250> (a org.eclipse.smarthome.automation.core.internal.RuleEngine)
        at org.eclipse.smarthome.automation.core.internal.RuleEngine.moduleTypeUpdated(RuleEngine.java:1265)
        at org.eclipse.smarthome.automation.core.internal.type.ModuleTypeManager.addingService(ModuleTypeManager.java:264)
        at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:942)
        at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:870)
        at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
        at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)
        at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:900)
        at com.prosyst.mbs.impl.framework.EventsManager.serviceChanged(EventsManager.java:1380)
        at com.prosyst.mbs.impl.framework.BundleContextImpl.registerService(BundleContextImpl.java:714)
        at com.prosyst.mbs.impl.framework.BundleContextImpl.registerService(BundleContextImpl.java:672)
        at com.prosyst.mbs.impl.framework.BundleContextImpl.registerService(BundleContextImpl.java:680)
        at com.prosyst.mbs.impl.services.automation.core.provider.ModuleTypeAdminResourceBundleProvider.run(ModuleTypeAdminResourceBundleProvider.java:221)
        - locked <0x00000007ac6846a0> (a java.util.HashMap)
        at com.prosyst.util.impl.tpt.threadpool.ExecutorImpl.run(ExecutorImpl.java:200)

Found 1 deadlock.
```

### CODE REVIEW

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 1]
```diff
@@ -114,7 +114,8 @@
 
     /**
      * {@link Map} of rule's id to corresponding {@link RuleEngineCallback}s. For each {@link Rule} there is one and
-     * only one rule callback.
+     * only
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: Remove new line

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 2]
```diff
@@ -181,10 +182,11 @@
     private ManagedRuleProvider managedRuleProvider;
 
     /**
-     * Constructor of {@link RuleEngine}. It initializes the logger and starts
-     * tracker for {@link ModuleHandlerFactory} services.
+     * Constructor of {@link RuleEngine}. It initializes the logger and starts tracker for {@link ModuleHandlerFactory}
+     * services.
      *
-     * @param bc {@link BundleContext} used for tracker registration and rule engine logger creation.
+     * @param bc
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: Any reason to insert the line break after the argument name?

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 3]
```diff
@@ -284,11 +295,14 @@ private void validateModules(List<Module> modules) {
 
     /**
      * Utility method which checks for existence of the rule with passed UID or create an unique id when the parameter
-     * is not passed
+     * is
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: New line

**Comment by maggu2810** [comment]
It seems something changed the JavaDoc a little bit.

I understand the code parts the synchronized has been moved to just that block that is need to be synchronized.

In "RuleRegistryImpl" you removed the synchronized completely. Is this correct and error-free (the iterations etc.)?

**Comment by danchom** [comment]
The RuleRegistry must not be synchronized, it calls methods from RuleEngine where the synchronization has to be done. If the registry methods are synchronized they will get the lock and entire rule engine's methods will be called in sync block.

**Pull request merged**